pkg/klusterlet/clusterclaim/labelclusterclaim_controller.go

```diff
@@ -0,0 +1,139 @@
+	clusterv1alpha1 "open-cluster-management.io/api/cluster/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func NewLableClusterClaimReconicler(
+	log logr.Logger,
+	clusterName string,
+	hubClient client.Client,
+	clusterClient clusterclientset.Interface,
+	clusterClaimLister clusterv1alpha1lister.ClusterClaimLister) (*LableClusterClaimReconicler, error) {
+	hubManaged, err := labels.NewRequirement(labelHubManaged, selection.Exists, nil)
+	if err != nil {
+		return nil, err
+	}
+	customizedOnly, err := labels.NewRequirement(labelCustomizedOnly, selection.Exists, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &LableClusterClaimReconicler{
+		log:                log,
+		clusterName:        clusterName,
+		hubClient:          hubClient,
+		clusterClient:      clusterClient,
+		clusterClaimLister: clusterClaimLister,
+		selector:           labels.NewSelector().Add(*hubManaged).Add(*customizedOnly),
+	}, nil
+}
+
+type LableClusterClaimReconicler struct {
+	log                logr.Logger
+	clusterName        string
+	hubClient          client.Client
+	clusterClient      clusterclientset.Interface
+	clusterClaimLister clusterv1alpha1lister.ClusterClaimLister
+
+	// selector filter the claims that should be reconciled and list out.
+	selector labels.Selector
+}
+
+func (r *LableClusterClaimReconicler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var err error
+	// Get Claims from managed cluster info
+	expectClaims, err := syncLabelsToClaims(r.hubClient, r.clusterName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Create/Update claims.
+	errs := []error{}
+	for _, claim := range expectClaims {
+		if err := createOrUpdateClusterClaim(ctx, r.clusterClient, claim, nil); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) != 0 {
+		return ctrl.Result{}, utils.NewMultiLineAggregate(errs)
+	}
+
+	// List existing claims.
+	existClusterClaims, err := r.clusterClaimLister.List(r.selector)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Clean unexpected ones.
+	return ctrl.Result{}, cleanClusterClaims(ctx, r.clusterClient, existClusterClaims, expectClaims)
+}
+
+func syncLabelsToClaims(hubClient client.Client, clusterName string) ([]*clusterv1alpha1.ClusterClaim, error) {
+	var claims []*clusterv1alpha1.ClusterClaim
+	request := types.NamespacedName{Namespace: clusterName, Name: clusterName}
+	clusterInfo := &clusterv1beta1.ManagedClusterInfo{}
+	err := hubClient.Get(context.TODO(), request, clusterInfo)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return claims, nil
+		}
+		return claims, err
+	}
+
+	// do not create claim if the label is managed by ACM.
+	// if the label format is aaa/bbb, the name of claim will be bbb.aaa.
+	// Besides, "_" and "/" in the label name will be replaced with "-" and "." respectively.
+	for label, value := range clusterInfo.Labels {
```

Comment:
we need sync the labels of managedCluster not managedClusterInfo, although the labels of managedcluster are included in the labels of managedClusterInfo.
but if meet error, the cluster labels are not synced to managedClusterInfo?

Reply:
This function is a copy from the old code with no change: https://github.com/stolostron/multicloud-operators-foundation/blob/111059e6a651503d3b6c89e9c62cfe65be5718f9/pkg/klusterlet/clusterclaim/clusterclaimer.go#L715

Agree to change to sync from managedcluster,  how about leaving a todo and fixing this with another refactor PR?